### ISSUE TITLE
Reconsider handling of `--`

### ISSUE THREAD

**Issue by rgov** [75lb/command-line-args#83]
The `--` argument has a well-defined meaning. `getopt(3)`:

> The interpretation of options in the argument list may be cancelled by the option `--` (double dash) which causes `getopt()` to signal the end of argument processing and return -1.

Python's `argparse` also [recognizes](https://github.com/python/cpython/blob/3.6/Lib/argparse.py#L1798) the double dash to mean that there are no more optional arguments, everything afterwards should be taken as a positional argument.

This project seems to avoid directly handling `--` itself and instead just stop parsing on the first unknown option, which means it will stop on `--` and then collect everything after as unparsed.

Since `--` is well-defined, the behavior should be to handle it directly. Everything after should only be interpreted as a positional argument / default option value.

**Comment by rgov** [comment]
Relatedly, `stopAtFirstUnknown` should probably be the default.

**Comment by 75lb** [comment]
> But then `--one a -- b c` is effectively treated the same as `--one -a --asdfasdf b c` which isn't desirable.

We can't guarantee it is not desirable, that's why this lib puts everything without an option definition into the `_unknown` property. This lets the developer decide what to do next. Some users of this lib only partially parse a few well-defined options from `argv` then pass the remaining `_unknown` into a separate third-party script which understands them. We can't make assumptions whether args received after the first unknown are useful or not.

**Comment by 75lb** [comment]
> stopAtFirstUnknown should probably be the default.

Yeah, i considered that too. But the current default is "strict mode" so changing it to `stopAtFirstUnknown` would make the default behaviour permissive - not something users have cried out for.

**Comment by 75lb** [comment]
The design goal here is to empower the developer to decide what `--` means, not dictate it.

**Comment by rgov** [comment]
So why is `--` a forbidden option?

**Comment by 75lb** [comment]
Have you read this? It's is an example how to implement `--` style behaviour.
https://github.com/75lb/command-line-args/wiki/Terminate-parsing-with-a-double-hypen

Could you give me an example of a command you want to run which uses `--` that is not currently possible with command-line-args?

**Comment by rgov** [comment]
It's not that it's not *possible*. After all, as I said, I can always take the `_unknown` array and do the argument parsing myself. It's that I'm using a library that should "just work" and handle conventional behavior for me.

Almost every standard UNIX tool understands `--`. It's a best practice if you are building command line argument arrays from user input, for instance, because the user's input might begin with a `-` inadvertently or even maliciously. When typing at the command line, most users skip it, but it's useful for `grep`, for instance, so that the pattern string doesn't get interpreted as an option.

The wiki page doesn't seem to give good solutions. Imagine I am the end user of a Node script written using `command-line-args`.

Suppose the script uploads a list of files to my FTP server. I try to write a conventional command line like `upload-file --ftp server.com -- *`. The `*` here is a glob that makes a list of every file in the current directory, but what if one of these files' names starts with a `-`? So the smart thing to do is precede it with `--` so that none of those file names is interpreted as an option to `upload-file`, such as `--delete-existing-files` (remember, maybe I'm dealing with malicious files created by attackers).

Unbeknownst to me, the developer of this script has not taken the extra effort to properly implement `--`, and the library has failed to handle it also. So when I run my command, it fails that `--` is an unexpected argument.
---

**Method 1** does not solve the problem, and just offers a workaround. Though this lets me get around the problem, how would I know to use `--file=`? I'm not the author of the script in question, and I don't know anything about `command-line-args` or even that it's being used. I would have to dig into the script, find its option definitions, figure out which one is the "optional value" (assuming I somehow knew what this meant), and the find the corresponding option name, *and then* use the `--option=value` notation. That's a big leap I just don't expect even technical end users to make.

But it still doesn't work. My simple, common shell glob `*` doesn't put `--file=` before each filename. So I'd finally have to do some shell scripting to do this.

---

*Method 2* suggests to use partial parsing. How maddening it would be if `upload-file` did so! The behavior would be very hard to understand to the end user.

But it doesn't solve the malicious file name case, so it's not usable.

---

*Method 3* is the only, I think, working solution, but it requires extra work for the developer.

* They need to set `stopAtFirstUnknown`.
* After parsing, they need to inspect the `_unknown` array.
* If the first element looks like an argument but is not `--`, then report the error to the user and stop. They cannot use the automatic error reporting provided by `command-line-parse`, but for consistency, they need to match its behavior closely.
* If the first element is `--`, then discard it.
* Concatenate the `_unknown` array with the default option's values.

This becomes much more complicated if positional arguments (#86) are supported. It's a lot of boilerplate to support a convention that is reasonable to expect the argument parser to handle.

---

It seems like the wiki should be updated, since it presents two inadequate solutions to the problem before arriving at the true solution that exists today, without any sample code. So even if some developer *did* know that `--` were not supported, and then somehow found this wiki page, read the whole thing rather than use the first problematic solution, and then decided to implement `--` themselves, they still wouldn't know the way to do it correctly.

**Comment by rgov** [comment]
I'll add, it makes sense for `--` to be handled in the parser and not after the fact because it changes the behavior of the parser *in the midst of parsing*. Semantically, it needs to be handled immediately when it appears, not after parsing is finished.
